1-Draft/RFCNNNN-ML-Based-Command-Line-Completion.md

```diff
@@ -0,0 +1,174 @@
+
+At this point it is not clear what the right machine learning model is to provide the best experience. It is expected to take many iterations of models and data processing before good and useful suggestions are provided. Customer privacy is critically important and so this feature will be opt-in since data collection is required to train the models. Collected data may be kept locally but more likely will be transferred to the cloud to create a larger data set for training. We must adhere to General Data Protection Regulation (GDPR) requirements to ensure that data is safe and in control of the customer.  
+
+The architecture will allow running components locally or in the cloud.  The cloud has tremendous processing power for training and running models, but latency may make locally running components more attractive in some cases.  
+
+Machine Learning is an inherently experimental process where different models and data features are examined, trained and evaluated. This RFC does not attempt to solve the machine learning and user experience problems, but instead provide an architecture to support experimentation of various Machine learning models and UI candidates.  
+
+### Goals
+
+* Support for both local and cloud based implementations of any component
+* Integrates with existing PowerShell command completion
+* Provides for privacy protection
+* Supports feature opt-in
+
+## Specification 
+
+### Modules
+#### Prediction Module  
+* Collects command line text used in PowerShell for model training
+* Returns prediction from submitted command line fragment
+* Creates suggestion success statistics based on user choice
+
+#### General Data Protection Regulation (GDPR) support Module
+These regulations apply to any personal data that is collected, and allows the user to:
+* View
+* Edit
+* Delete
+* Restrict
+* Export
+
+The mechanism for feeding the model with data and retrieving results will closely follow the existing command completion implementation.  The Prediction Module will implement APIs that will:  
+* Accept command line fragments with which to generate predictions
+* Provide prediction information including confidence level
+* Collect complete command lines processed by PowerShell (GDPR applied)
+* Generate success statistics based on user choices
+
+The user experience will be dependent on the PowerShell host and is outside the scope of this RFC. But hosts will need to be updated to take advantage of the new opt-in prediction feature. For example PowerShell shell PSReadline should integrate it into its tab completion and/or Ctrl+Space suggestion list display. VSCode should likewise integrate it into its tab completion and intellisense.  

```

Comment:
I think we probably need to provide a UX implementation because:
1. We don't know if the infrastructure would work unless we really consume the interfaces to implement a UX
2. We need a UX to attract users to opt in the feature.

Reply:
On this note - I have no intention of merging changes to `PSReadLine` that utilizes cloud data (sending or receiving) without **strong** feedback from users that they actually want this feature.

It doesn't matter if cloud data is opt-in - people forget when they try stuff out, and people forget that they shouldn't use secrets at the command line. It should be impossible for such secrets to leak unless the machine holding those secrets (in a history file or transcript) was compromised. If the secrets end up in the cloud, then a leak becomes possible.

Reply:
@daxian-dbw  Well yes, of course the UI needs to support this before it can be tested.  But I didn't want to impose UX requirements in this RFC because I feel a good UX will emerge through experimentation.  But if you have UX ideas please feel free to add language here and I'll include it in the PR (or it can be added later).  After all you are a co-author :)

@lzybkr Definitely.  Handling sensitive command line data is a huge concern.  That is why one idea is to only collect and train data locally, but I don't know how realistic this would be or what the perf impact would be.  Another idea would be to have PSReadline augment the prompt to clearly indicate the feature is active (if possible).

Reply:
@lzybkr I think the only change we would submit to PSReadline is rendering multiple suggestions.  The handling of cloud data would be in PowerShell.